caps/management/commands/process_scottish_reporting_data.py

```diff
@@ -0,0 +1,428 @@
+        if type_header not in targets_df.columns:
+            type_header = "Name of target"
+
+        for index, row in targets_df.iterrows():
+            point_type = row[type_header]
+            point_data = row["Target"]
+            point_units = row["Units"]
+            year = row["Target completion year"]
+            if point_data in [
+                "Please select from drop down box",
+                "Other (please specify in comments)",
+            ]:
+                continue
+            self.make_data_point(
+                point_type="target",
+                point_data=point_data,
+                sub_type=point_type,
+                units=point_units,
+                target_year=year,
+            )
+
+        return 1
+
+    def extract_projects_from_sheet(self, targets_df):
+        """
+        we don't replace the columns with our own as sadly they aren't always in the
+        same order so instead we have to do some comparison to work out which column to use.
+        Note that some of them are the same, only with spaces at the end.
+        """
+        measurement_columns = [
+            "Are these savings figures estimated or actual? ",
+            "Are these savings figures estimated or actual?",
+            "Savings figures are estimated or actual ",
+            "Savings figures are estimated or actual",
+        ]
+
+        measurement = "Savings figures are estimated or actual"
+        for col in measurement_columns:
+            if col in targets_df.columns:
+                measurement = col
+
+        for col in [
+            "Primary fuel/emission source saved",
+            measurement,
+            "First full year of CO2e savings",
+        ]:
+            targets_df[col] = targets_df[col].replace(
+                "Please select from drop down box", ""
+            )
+
+        for index, row in targets_df.iterrows():
+            point_type = row["Project name"].strip()
+            point_data = row["Estimated carbon savings per year (tCO2e/annum)"]
+            funding_source = row["Funding source"]
+            if type(funding_source) == str:
+                funding_source = funding_source.strip()
+            self.make_data_point(
+                point_type="project",
+                point_data=point_data,
+                sub_type=point_type,
+                lifetime=row["Project lifetime (years)"],
+                cost=row["Capital cost (£)"],
+                funding_source=funding_source,
+                emission_source=row["Primary fuel/emission source saved"],
+                annual_cost=row["Operational cost (£/annum)"],
+                annual_savings=row["Estimated costs savings (£/annum)"],
+                measurement=row[measurement],
+                savings_start=row["First full year of CO2e savings"],
+                comments=row["Comments"],
+            )
+
+        return 1
+
+    def extract_generation_from_sheet(self, df):
+        if len(df.columns) == 3:
```

Comment:
What leads to this being needed?

Reply:
sigh, I think I did this before going away for a bit so only did half of it. I've done the rest now in a slightly better way with an explanation of what is going on.